### PR TITLE
[feature-wip][MTMV] The schedule framework for the MTMV

### PR DESCRIPTION
**Pull request by SaintBacchus**
# Proposed changes

Issue Number: #13146

## Problem summary

Describe in issue.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 1]
```diff
@@ -1792,4 +1792,19 @@ public class Config extends ConfigBase {
...
     public static int statistic_task_scheduler_execution_interval_ms = 60 * 1000;
+
+    /*
+     * mtmv scheduler framework is still under dev, remote this config when it is graduate.
+     */
+    @ConfField(mutable = true)
+    public static boolean enable_mtmv_scheduler_framework = false;
+
+    @ConfField(mutable = true, masterOnly = true)
+    public static int running_mtmv_scheduler_task_size = 1000;
```
Comment (yangzhg) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1803]:
```suggestion
    public static int max_running_mtmv_scheduler_task_num = 1000;
```

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 2]
```diff
@@ -1792,4 +1792,19 @@ public class Config extends ConfigBase {
...
+
+    @ConfField(mutable = true, masterOnly = true)
+    public static int pending_mtmv_scheduler_task_size = 1000;
```
Comment (yangzhg) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1806]:
```suggestion
    public static int max_pending_mtmv_scheduler_task_num = 100;
```

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/metadata/ChangeMtmvTask.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import org.apache.doris.mtmv.MtmvUtils.TaskState;
+import org.apache.doris.persist.gson.GsonUtils;
+
+import com.google.gson.annotations.SerializedName;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+public class ChangeMtmvTask implements Writable {
```
Comment (yangzhg) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/metadata/ChangeMtmvTask.java:31]:
```suggestion
public class AlterMtmvTask implements Writable {
```

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 40.06 seconds
 load time: 597 seconds
 storage size: 17154768637 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221019205855_clickbench_pr_31435.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.87 seconds
 load time: 567 seconds
 storage size: 17154699298 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221021123300_clickbench_pr_32323.html

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 4]
```diff
@@ -432,6 +433,8 @@ public class Env {
 
     private PolicyMgr policyMgr;
 
+    private MtmvJobManager mtmvJobManager;
```
Comment (adonis0147) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:436]: `MTMV` is short for *Multiple Table Materialized Table*. I think to use `MTMV` (all capital characters) is better.

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/metadata/MTMVTask.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    public String getUser() {
+        return user;
+    }
+
+    public void setUser(String user) {
+        this.user = user;
+    }
+
+    public String getDefinition() {
```
Comment (adonis0147) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/metadata/MTMVTask.java:129]: `getDefinition` -> `getQuery`

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVJobFactory.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.concurrent.TimeUnit;
+
+public class MTMVJobFactory {
+    public static MTMVJob buildJob(MaterializedView materializedView) {
+        MTMVJob job = new MTMVJob(materializedView.getName());
+        JobSchedule jobSchedule = new JobSchedule(System.currentTimeMillis() / 1000, 1, TimeUnit.MINUTES);
+        job.setSchedule(jobSchedule);
+        job.setTriggerMode(TriggerMode.PERIODICAL);
+        job.setDbName("fake");
+        job.setDefinition("select * from fake");
```
Comment (adonis0147) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVJobFactory.java:34]: `setDefinition` -> `setQuery`

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 7]
```diff
@@ -2091,6 +2100,12 @@ private void createOlapTable(Database db, CreateTableStmt stmt) throws UserExcep
 
             throw e;
         }
+
+        // TODO: impl the real logic of create multi-table MaterializedView here.
+        if (olapTable instanceof MaterializedView && Config.enable_mtmv_scheduler_framework) {
+            Env.getCurrentEnv().getMtmvJobManager().createJob(MTMVJobFactory.buildJob((MaterializedView) olapTable),
```
Comment (adonis0147) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:2106]: `getMtmvJobManager` -> `getMTMVJobManager`

File: fe/fe-core/src/main/java/org/apache/doris/persist/EditLog.java [hunk 8]
```diff
@@ -884,6 +890,36 @@ public static void loadJournal(Env env, JournalEntity journal) {
                     env.getLoadManager().replayCleanLabel(log);
                     break;
                 }
+                case OperationType.OP_CREATE_MTMV_JOB: {
+                    final MTMVJob job = (MTMVJob) journal.getData();
+                    env.getMtmvJobManager().replayCreateJob(job);
```
Comment (adonis0147) [inline fe/fe-core/src/main/java/org/apache/doris/persist/EditLog.java:895]: `getMtmvJobManager` -> `getMTMVJobManager`

**Review by adonis0147 — APPROVED** [review]
LGTM

**Pull request merged**